lib/predicate/registry.js

```diff
@@ -0,0 +1,367 @@
+
+function getPredicate(registry, runParams, callback) {
+    const ve =
+        validate(runParamsSpec, runParams, new PredicateError(), runParams);
+    if (ve.hasValues()) {
+        return callback(ve);
+    }
+
+    const key = createKey(runParams);
+    const preds = registry.predicates[key];
+
+    if (!preds) {
+        return callback(ve.push(`No predicate exists for "${key}"`));
+    }
+
+    const k = runParams.key;
+    const klen = k.length;
+    let maxp = -1;
+    let fn = null;
+
+    // Linear search - for now we don't anticipate
+    // having too many prefixes for the same bucket.
+    Object.keys(preds).forEach(p => {
+        const plen = p.length;
+        if (plen > klen) {
+            return;
+        }
+        if (plen > maxp && k.indexOf(p) === 0) {
+            maxp = plen;
+            fn = preds[p];
+        }
+    });
+
+    if (!fn) {
+        return callback(ve.push(`No predicate exists for "${key}"`));
+    }
+    return callback(null, fn);
+}
+
+function runWrapped(context, callback) {
+    const { passthrough, event, fn } = context;
+    // NOTE: Domains are used to prevent exceptions from
+    // killing the worker process (which is not possible with
+    // promises). See tests for details.
+    const d = domain.create();
+    d.on('error', err => {
+        d.exit();
+        return callback(err);
+    });
+    d.run(() => fn(event, (err, res, body) => {
+        d.exit();
+        if (err) {
+            return callback(new PredicateError(err));
+        }
+        if (!body) {
+            return callback(null, res, passthrough);
+        }
+        // The user-supplied predicated transformed the object.
+        const output = new stream.PassThrough();
+        if (body instanceof stream.Readable) {
+            const chunks = [];
+            body.on('data', c => {
+                chunks.push(c);
+                output.write(c);
+            });
+            body.on('end', () => {
+                output.parsedContentLength =
+                            Buffer.concat(chunks).length;
+                output.end();
+            });
+        } else {
+            output.parsedContentLength = Buffer.byteLength(body);
+            output.end(body);
+        }
+        return callback(null, res, output);
+    }));
+}
+
+function runPredicate(runParams, fn, callback) {
+    const { request } = runParams;
+    const chunks = [];
+    const passthrough = new stream.PassThrough();
+    passthrough.headers = request.headers; // TODO: do we need anything else??
+    passthrough.parsedContentLength = request.parsedContentLength;
+
+    request.on('data', c => {
+        chunks.push(c);
+        passthrough.write(c);
+    });
+    request.on('error', err => {
+        passthrough.end();
+        return callback(err);
+    });
+    request.on('end', () => {
+        passthrough.end();
+        return runWrapped({
+            passthrough,
+            event: makePredicateEvent(runParams, Buffer.concat(chunks)),
```

Comment:
So the user's `data` api would be a stream? I was trying to keep this super simple for end users, but I agree, _if_ they want the data (and they might not care since a key use case for this functionality might be metadata validation), they can should use a stream.